### PR TITLE
Allow to delete a staging project via API

### PR DESCRIPTION
**Pull request by saraycp**
Also documents it in apidocs.

### CODE REVIEW

**Comment by saraycp** [comment]
[Allow to Destroy/Remove a Staging Project via the API](https://trello.com/c/2zIIt0hB/1076-allow-to-destroy-remove-a-staging-project-via-the-api)

**Comment by DavidKang** [comment]
Looks good, but I think we need to provide a way to delete multiples staging project instead of doing it one by one.

File: docs/api/api/api.txt [hunk 1]
```diff
@@ -1947,6 +1947,10 @@ GET /staging/<staging_workflow_project>/staging_projects/<staging_project>
   Get the overall state of a staging project
 XmlResult: staging_project
 
+DELETE /staging/<staging_workflow_project>/staging_projects/<staging_project>
```
Comment (coolo) [inline docs/api/api/api.txt]: why would I delete staging projects differently than regular projects? I guess I need a little context here (for the docs)

**Comment by coolo** [comment]
> Looks good, but I think we need to provide a way to delete multiples staging project instead of doing it one by one.

For what user story?

File: docs/api/api/api.txt (hunk 1 above)
Comment (saraycp) [reply docs/api/api/api.txt]: @coolo I don't know If I understood :thinking: Do you propose to have a similar route or use exactly the same for both regular and staging projects?

**Comment by saraycp** [comment]
> >  Looks good, but I think we need to provide a way to delete multiples staging project instead of doing it one by one.
> 
> For what user story?

@coolo in the new UI we do it this way, we have a list of staging projects that we can select with checkboxes and then delete all of them. We should do the same for the API. Don't you think it is necessary?

File: docs/api/api/api.txt (hunk 1 above)
Comment (coolo) [reply docs/api/api/api.txt]: Well, I propose you clarify in the documentation why it's necessary to have an extra route and what's breaking if I use the /source route to delete the project.

Comment (coolo) [reply docs/api/api/api.txt]: But I'm not even sure this is the case.

**Comment by coolo** [comment]
> @coolo in the new UI we do it this way, we have a list of staging projects that we can select with checkboxes and then delete all of them. We should do the same for the API. Don't you think it is necessary?

I wonder how you envision the use of this API - if there is need to delete multiple projects from a client, that client can very well run a loop.

**Comment by saraycp** [comment]
Ok, we have a few things to clarify.

First, we don't have to do a multiple deletion, this PR is about deleting a single staging project (SP). We got confused with the staging workflow (SW) deletion where we decide what to do with its SPs.

Second, do we need a new route to remove a SP if it is a project after all? Apparently not, but...  deleting with `/source/<project>` or `/staging/<staging_workflow_project>/staging_projects/<staging_project>` is slightly different. We have realized that only the SP class calls `after_destroy :update_staging_workflow_on_backend` ([line 218](https://github.com/openSUSE/open-build-service/blob/0ae5b0806ecbc1a84f1e222500c26e82a8ee5f1f/src/api/app/models/staging/staging_project.rb#L218)) that updates the SW xml. After deletion, it changes this

```
<workflow project=\"test_project\" managers=\"heroes\">
  <staging_project name=\"test_project:Staging:A\"/>
</workflow>
```
to this
```
<workflow project=\"test_project\" managers=\"heroes\"></workflow>
```

We can move the code that updates the SW on backend to the Project model, what I think is a bit untidy from the coding point of view. Or we can keep both routes, what I think is a bit risky if someone deletes the SP using the old route `/source/<project>` by accident, we will have an inconsistent SW on backend.

@coolo @hennevogel @DavidKang what do you think?

File: src/api/config/routes.rb [hunk 2]
```diff
@@ -762,7 +762,7 @@ def self.public_or_about_path?(request)
 
   # StagingWorkflow API
   resources :staging, only: [], param: 'main_project_name' do
-    resources :staging_projects, only: [:index, :show], controller: 'staging/staging_projects', param: :name do
+    resources :staging_projects, only: [:index, :show, :destroy], controller: 'staging/staging_projects', param: :name do
```
Comment (Ana06) [inline src/api/config/routes.rb]: merge conflict :see_no_evil: https://github.com/openSUSE/open-build-service/pull/7161/commits/0dc1b55cf2866a6c39d7e2714d78d71dd4d811ff

File: docs/api/api/api.txt (hunk 1 above)
Comment (Ana06) [reply docs/api/api/api.txt]: actually, what prevent you to use the normal project route? :thinking:

**Comment by saraycp** [comment]
> actually, what prevent you to use the normal project route? 

@Ana06 nothing, like it is you can use both. But there is a slightly difference as I mentioned before.

**Comment by coolo** [comment]
> @coolo @hennevogel @DavidKang what do you think?
OK, let's talk webui first. https://build.opensuse.org/project/show/TestLinux:Staging:A is a (staging) project and there is a 'Delete Project' button on it. Does this delete a project or a staging project? 

From my point of view: deleting a project, no matter if it's a maintenance project, a staging project, an incident project or a ... - it should always trigger the right things when deleted (or be blocked, which might be the right thing too).

File: docs/api/api/api.txt (hunk 1 above)
Comment (vpereira) [reply docs/api/api/api.txt]: are you able to delete multiple projects via the ```/source``` route @coolo ?

**Comment by saraycp** [comment]
> OK, let's talk webui first. https://build.opensuse.org/project/show/TestLinux:Staging:A is a (staging) project and there is a 'Delete Project' button on it. Does this delete a project or a staging project?

This deletes a regular project.

> From my point of view: deleting a project, no matter if it's a maintenance project, a staging project, an incident project or a ... - it should always trigger the right things when deleted (or be blocked, which might be the right thing too).

After some discussion we came to the same conclusion. So we are going to use only one endpoint to delete via API: `/source/<project>`. And we are going to move the callbacks specifically defined for staging projects to project model.

Superseded by: #7172

**Pull request closed without merging**